File: Solutions/CSharp/The-Chamber-of-Echoes.cs
public class ChamberOfEchoes
{
    static List<int> echoes = new List<int> { 3, 6, 9, 12 };
    static List<int> memories = new List<int>();

    // Predicts the next number in an arithmetic sequence
    // by calculating the common difference and applying it to the last element
    static int PredictNext(List<int> echoes)
    {
        int commonDifference = CalculateCommonDifference(echoes);
        int predictedValue = echoes[echoes.Count - 1] + commonDifference;
        
        StoreSequenceInMemories(echoes, predictedValue);
        
        return predictedValue;
    }

    static int CalculateCommonDifference(List<int> sequence)
    {
        return sequence[1] - sequence[0];
    }

    static void StoreSequenceInMemories(List<int> sequence, int predictedValue)
    {
        memories.AddRange(sequence);
        memories.Add(predictedValue);
    }

    public static void Run()
    {
        Console.WriteLine(PredictNext(echoes));
    }
}

File: Solutions/Ruby/The-Chamber-Of-Echoes.rb
# The Chamber of Echoes
# Adventure: The Chamber of Echoes
#
# Predicts the next number in a sequence by computing the constant difference
# between consecutive terms, then adding it to the last term.
# The full sequence (including the predicted value) is stored in memories.

echoes  = [3, 6, 9, 12]
memories = []

# Predict the next value in the sequence and store the extended sequence
def predict_next(echoes, memories)
  difference = echoes[1] - echoes[0]
  next_value = echoes.last + difference
  memories.concat(echoes + [next_value])
  next_value
end

puts predict_next(echoes, memories)
# => 15
